src/eflambe_server.erl

```diff
@@ -47,102 +58,176 @@
+            {ok, #state{module = Module, max_calls = MaxCalls, options = FinalOptions}}
+    end.
 
 -spec handle_call(Request :: any(), from(), state()) ->
                                   {reply, Reply :: any(), state()} |
-                                  {reply, Reply :: any(), state(), timeout()}.
+                                  {reply, Reply :: any(), state(), {continue, finish}}.
 
-handle_call({start_trace, Id, CallLimit, Options}, {FromPid, _}, State) ->
-    TracerOptions = [{pid, FromPid}|Options],
-    case get_trace_by_id(State, Id) of
-        undefined ->
-            % Create new trace, spawn a tracer for the trace
-            {ok, TracerPid} = eflambe_tracer:start_link(TracerOptions),
-            UpdatedTrace = #trace{
-                              id = Id,
-                              max_calls = CallLimit,
-                              calls = 1,
-                              options = Options,
-                              running = true,
-                              tracer = TracerPid
-                             },
-            {reply, {ok, Id, true, TracerPid}, put_trace(State, UpdatedTrace)};
-        #trace{max_calls = MaxCalls, calls = Calls, tracer = TracerPid, running = false}
-          when Calls =:= MaxCalls ->
-            {reply, {ok, Id, false, TracerPid}, State};
-        #trace{max_calls = MaxCalls, calls = Calls, tracer = TracerPid, running = true}
-          when Calls =:= MaxCalls ->
-            {reply, {ok, Id, false, TracerPid}, State};
-        #trace{calls = Calls, running = false} = Trace ->
-            % Increment existing trace
+handle_call(start_trace, {FromPid, _}, #state{max_calls = MaxCalls, calls = Calls,
+                                              options = Options} = State) ->
+    NoTracesRemaining = (MaxCalls =:= Calls),
+
+    case {get_pid_trace(State, FromPid), NoTracesRemaining} of
+        {_, true} ->
+            {reply, {ok, false}, State};
+        {undefined, false} ->
+            % Increment number of calls
             NewCalls = Calls + 1,
-            % Create new trace, spawn a tracer for the trace
+
+            % Start a new tracer
+            TracerOptions = [{pid, FromPid}|Options],
             {ok, TracerPid} = eflambe_tracer:start_link(TracerOptions),
 
-            % Update number of calls
-            UpdatedTrace = Trace#trace{calls = NewCalls, running = true, tracer = TracerPid},
-            NewState = update_trace(State, Id, UpdatedTrace),
-            {reply, {ok, Id, true, TracerPid}, NewState};
-        #trace{options = Options, running = true, tracer = TracerPid} ->
-            {reply, {ok, Id, false, TracerPid}, State}
+            % Update state
+            NewPidTrace = #pid_trace{ pid = FromPid, tracer_pid = TracerPid},
+            NewState = put_pid_trace(State#state{calls = NewCalls}, NewPidTrace),
+
+            % Create new trace record, start tracing function
+            start_erlang_trace(FromPid, TracerPid),
```

Comment:
Move this to `start_capture_trace/1` function so it's invoked from the calling process and not from the `gen_server`. This will result in the `gen_server` call itself not being recorded, which is what we want.